File: daily/437-Path-sum-iii.go
package daily

func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}

	return dfsPathSum(root, 0, targetSum) + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}

func dfsPathSum(node *TreeNode, currentSum int, targetSum int) int {
	if node == nil {
		return 0
	}

	currentSum += node.Val
	count := 0
	if currentSum == targetSum {
		count = 1
	}
	count += dfsPathSum(node.Left, currentSum, targetSum)
	count += dfsPathSum(node.Right, currentSum, targetSum)
	return count
}

File: daily/Binary Tree/437-Path-sum-iii.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root:
            return 0
        
        def dfs(node, currentSum):
            if not node:
                return 0
            
            currentSum += node.val
            count = 1 if currentSum == targetSum else 0
            left = dfs(node.left, currentSum)
            right = dfs(node.right, currentSum)
            return count + left + right
        return (dfs(root, 0) + self.pathSum(root.left, targetSum) + self.pathSum(root.right, targetSum))